src/par_iter/chain.rs

```diff
@@ -199,14 +206,58 @@ impl<A, B> Producer for ChainProducer<A, B>
     }
 }
 
-impl<A, B> IntoIterator for ChainProducer<A, B>
-    where A: Producer,
-          B: Producer<Item = A::Item>
+/// ////////////////////////////////////////////////////////////////////////
+/// Wrapper for Chain to implement ExactSizeIterator
+
+pub struct Chain<A, B> {
+    chain: iter::Chain<A, B>,
+    len: usize,
+}
+
+impl<A, B> Chain<A, B> {
+    fn new(a: A, b: B) -> Chain<A, B>
+        where A: ExactSizeIterator,
+              B: ExactSizeIterator<Item=A::Item>
+    {
+        let len = a.len() + b.len();
```

Comment:
I think it's fine to check, but if there's an overflow I expect we'll first have a problem in `ExactParallelIterator::len()`.  So let's check everywhere it matters, if we're going to check at all.

I wonder here though, do you really need to keep track the length at all?  You're paying a (small) computational cost to update a value that's only occasionally read.  Plus, I don't think you need it anyway, because `ExactSizeIterator::len()` has a default implementation that just checks and uses the value from `Iterator::size_hint()`, which should work fine here if you pass it through `Chain` wrapper.  In fact, the docs say:

> When implementing an `ExactSizeIterator`, You must also implement `Iterator`. When doing so, the implementation of `size_hint()` _must_ return the exact size of the iterator.

... which implies that you can't leave `size_hint()` to the default `(0, None)`.